### PR TITLE
Encrypt offline db

### PR DESCRIPTION
**Pull request by johnbotris**
TODO (archived):
* It's possible that the user might be prompted to unlock their keychain twice when unlocking existing credentials but without already having a database key, if the keychain is set to never remain unlocked. We should decide if this is acceptable. (it would only ever happen once at the most for any user) 
* If the database or the database key are somehow corrupted, then the app will be broken until the user manually deletes the database. We should delete the file automatically in this case.
    * This ties into part of the acceptance criteria for https://github.com/tutao/tutanota/issues/3812, in that when a user deletes their credentials and an offline database exists, further attempts to save the same credentials with a new session will be broken, as the original database key will be lost but the database itself remains.

* automation of building OpenSSL for windows needs to be setup. cross compilation does not work for our use-case, so we have to install perl and NASM on the windows build machine in order to do it natively (i have notes which can be referred to). This could be considered another task since we already have our own build of openssl for each platform. We also want to setup a CI build for OpenSSL for iOS, maybe they can be combined.
* a pipeline for generating our own SQLCipher amalgamation would also be useful for future updates (it can work as documentation on how to build it as well)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,90 +1,90 @@
 {
-	"name": "tutanota",
+  "name": "tutanota",
```
Comment (charlag) [inline package.json:2]: package.json has it's own weird whitespace situation because npm

File: OpenSSL.Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+					publish("mac")
+// 					publish("win")
+				}
+            }
+        }
+    }
+}
+
+def buildOpenssl(Map params) {
+	sh "rm -rf openssl" // Jenkins doesn't seem to use a clean workspace each time
```
Comment (charlag) [inline OpenSSL.Jenkinsfile:55]: you can set a flag to force wipe each time

File: OpenSSL.Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	sh "git clone git://git.openssl.org/openssl.git"
```
[the remark this answers is not in the document]
Comment (johnbotris) [reply OpenSSL.Jenkinsfile:56]: builtin for which thing? the source code?

File: OpenSSL.Jenkinsfile (hunk 2 above)
Comment (johnbotris) [reply OpenSSL.Jenkinsfile:55]: do we use it somewhere else?

Comment (charlag) [reply OpenSSL.Jenkinsfile:55]: Yes, we do it at least for admin client, I would recommend always using that

File: src/api/worker/rest/CacheStorageProxy.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ * Create a proxy to a cache storage object.
+ * It will be uninitialized, and unusable until {@method LateInitializedCacheStorage.initializeCacheStorage} has been called on the returned object
+ * Once it is initialized, then it is safe to use
+ * @param factory A factory function to get a CacheStorage implementation when initialize is called
+ * @return {LateInitializedCacheStorage} The uninitialized proxy and a function to initialize it
+ */
+export function makeStorageProxy(factory: (args: StorageInitArgs) => Promise<CacheStorage>): LateInitializedCacheStorage {
+	let actualStorage: CacheStorage | null = null
+
+	const proxy = new Proxy({}, {
```
Comment (charlag) [inline src/api/worker/rest/CacheStorageProxy.ts:27]: oh I thought you wouldn't make an actual `Proxy` but just another impl with deletegation but I guess this also works?

Comment (johnbotris) [reply src/api/worker/rest/CacheStorageProxy.ts:27]: Is there a reason it wouldn't work?

Comment (charlag) [reply src/api/worker/rest/CacheStorageProxy.ts:27]: Idk, proxy gives unexpected result but I think it should be fine

File: src/misc/credentials/NativeCredentialsEncryption.ts [hunk 5]
```diff
@@ -43,28 +49,49 @@ export class NativeCredentialsEncryption implements CredentialsEncryption {
...
 	}
 
-	async decrypt(encryptedCredentials: PersistentCredentials): Promise<Credentials> {
-		const credentialsKey = await this._credentialsKeyProvider.getCredentialsKey()
-		const decryptedAccessToken = await this._deviceEncryptionFacade.decrypt(credentialsKey, base64ToUint8Array(encryptedCredentials.accessToken))
-			.catch(ofClass(CryptoError, (e) => {
+	async decrypt(encryptedCredentials: PersistentCredentials): Promise<CredentialsAndDatabaseKey> {
+		const credentialsKey = await this.credentialsKeyProvider.getCredentialsKey()
+
+		let accessToken,
```
Comment (charlag) [inline src/misc/credentials/NativeCredentialsEncryption.ts:59]: wtf

File: package.json [hunk 6]
```diff
@@ -33,7 +33,7 @@
 		"@types/qrcode-svg": "1.1.1",
 		"@types/systemjs": "6.1.1",
 		"@types/winreg": "1.2.31",
-		"better-sqlite3": "7.5.0",
+		"better-sqlite3": "https://github.com/tutao/better-sqlite3-sqlcipher.git#better-sqlcipher",
```
[the remark this answers is not in the document]
Comment (johnbotris) [reply package.json:36]: yes, but there are still some changes to be made in that repo, so it will come afterwards

**Comment by johnbotris** [comment]
TODO kinda:
* It's possible that the user might be prompted to unlock their keychain twice when unlocking existing credentials but without already having a database key, if they have the keychain is set to never remain unlocked. We should decide if this is acceptable. (it would only ever happen once at the most for any user) 
* If the database or the database key are somehow corrupted, then the app will be broken until the user manually deletes the database. We should delete the file automatically in this case.
    * This ties into part of the acceptance criteria for https://github.com/tutao/tutanota/issues/3812, in that when a user deletes their credentials and an offline database exists, further attempts to save the same credentials with a new session will be broken, as the original database key will be lost but the database itself remains.

* automation of building OpenSSL for windows needs to be setup. cross compilation does not work for our use-case, so we have to install perl and NASM on the windows build machine in order to do it natively (i have notes which can be referred to). This could be considered another task since we already have our own build of openssl for each platform. We also want to setup a CI build for OpenSSL for iOS, maybe they can be combined.
* a pipeline for generating our own SQLCipher amalgamation would also be useful for future updates (it can work as documentation on how to build it as well)

**Pull request merged**